### ISSUE TITLE
[1.17.1] forge:nbt ingredient fails to sync to the client

### ISSUE THREAD

**Issue by Mrbysco** [MinecraftForge/MinecraftForge#7933]
**Minecraft Version:** 1.17.1

**Forge Version:** 37.0.10

**Logs:** 
[latest.log](https://gist.github.com/Mrbysco/4e227fe2ea59f6d055b3d9b1c132131d)
[nbt_datapack.zip](https://github.com/MinecraftForge/MinecraftForge/files/6878764/nbt_datapack.zip)

**Steps to Reproduce:**
 1. Add a recipe to a mod that requires the `forge:nbt` ingredient
 2. Start a server and a client
 3. Connect the client to the server
 4. Observe how the client gets denied with a DecoderException

**Description of issue:**
The `forge:nbt` ingredient fails to sync to the client when connecting to a server.

I've provided a datapack to reproduce the issue.

**Comment by GeheimagentNr1** [comment]
I had the same issue and found, that there is a problem in the fromNetwork/toNetwork methods of the Ingredient class:
This is the 1.17.1 version of the Ingredient#toNetwork methods:
```java
   public final void toNetwork(FriendlyByteBuf p_43924_) {
      this.dissolve();
      if (!this.isVanilla()) {
         net.minecraftforge.common.crafting.CraftingHelper.write(p_43924_, this);
         return;
      }
      p_43924_.writeCollection(Arrays.asList(this.itemStacks), FriendlyByteBuf::writeItem);
   }
```
This is the 1.17.1 version of the Ingredient#fromNetwork method:
```java
   public static Ingredient fromNetwork(FriendlyByteBuf p_43941_) {
      return fromValues(p_43941_.readList(FriendlyByteBuf::readItem).stream().map(Ingredient.ItemValue::new));
   }
```
In version 1.16.5 the Ingredient#fromNetwork method looked like that:
```java
   public static Ingredient fromNetwork(PacketBuffer p_199566_0_) {
      int i = p_199566_0_.readVarInt();
      if (i == -1) {
         return net.minecraftforge.common.crafting.CraftingHelper.getIngredient(p_199566_0_.readResourceLocation(), p_199566_0_);
      }
      return fromValues(Stream.generate(() -> {
         return new Ingredient.SingleItemList(p_199566_0_.readItem());
      }).limit((long)i));
   }
```
The handling for custom Ingredient is missing in the 1.17.1 version of the Ingredient#fromNetwork method.

I don't know how to fix this, but at least I found it and shouldn't be hard to fix.

**Comment by sciwhiz12** [comment]
This seems to have been fixed by #7980.
